### ISSUE TITLE
There seems to be a problem with whether the aiMessageFrom parsing function is used

### ISSUE THREAD

**Issue by nn200433** [langchain4j/langchain4j-examples#28]
chatglm responds as follows:
```json
{
    "model": "chatglm3",
    "object": "chat.completion",
    "choices": [
        {
            "index": 0,
            "message": {
                "role": "assistant",
                "content": "get_weather\n ```python\ntool_call(city_name='北京')\n```",
                "name": null,
                "function_call": {
                    "name": "get_weather",
                    "arguments": "{\"city_name\": \"北京\"}"
                }
            },
            "finish_reason": "function_call"
        }
    ],
    "created": 1702883142,
    "usage": {
        "prompt_tokens": 285,
        "total_tokens": 307,
        "completion_tokens": 22
    }
}
```

Cannot be parsed using the following code：
```java
public static AiMessage aiMessageFrom(ChatCompletionResponse response) {
    if (response.content() != null) {
        return aiMessage(response.content());
    } else {
        FunctionCall functionCall = response.choices().get(0).message().functionCall();

        ToolExecutionRequest toolExecutionRequest = ToolExecutionRequest.builder()
                .name(functionCall.name())
                .arguments(functionCall.arguments())
                .build();

        return aiMessage(toolExecutionRequest);
    }
}
````

**Comment by langchain4j** [comment]
@nn200433 so you mean we should check if there is function first?

**Comment by langchain4j** [comment]
this was [fixed](https://github.com/langchain4j/langchain4j/blob/38049a197bac084b9f980a7edaf9619e0115403d/langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/InternalOpenAiHelper.java#L53), will be available in 0.25.0

**Comment by nn200433** [comment]
> @nn200433 so you mean we should check if there is function first?

First of all, thanks for your reply！My mother tongue is not English, so English is very poor, please forgive me.

Yes, I mean such a request response, which cannot recognize that the function call was triggered.

this is my modify:
```java
public static AiMessage aiMessageFrom(ChatCompletionResponse response) {
    final Message      message      = response.choices().get(0).message();
    final FunctionCall functionCall = message.functionCall();
    if (functionCall != null) {
        ToolExecutionRequest toolExecutionRequest = ToolExecutionRequest.builder()
                .name(functionCall.name())
                .arguments(functionCall.arguments())
                .build();
        // Prevents the content of the history message list some one from being null
        // error request message body: [{"role":"system","content":"example content..."},{"role":"user","content":"example content..."},{"role":"assistant","content":"example content..."},{"role":"user","content":"帮我查下\n预定编号：123-457\n姓名：Klaus Heisler"},{"role":"assistant","content":null,"function_call":{"name":"getBookingDetails","arguments":"{\"bookingNumber\": \"123-457\", \"customerName\": \"Klaus Heisler\"}"}},{"role":"function","content":"{\n  \"bookingNumber\": \"123-457\",\n  \"customer\": {\n    \"name\": \"Klaus Heisler\"\n  }\n}","name":"getBookingDetails"}]
        // java.lang.RuntimeException: dev.ai4j.openai4j.OpenAiHttpException: {"detail":[{"type":"string_type","loc":["body","messages",4,"content"],"msg":"Input should be a valid string","input":null,"url":"https://errors.pydantic.dev/2.5/v/string_type"}]}
        return new AiMessage(response.content(), toolExecutionRequest);
    } else {
        return aiMessage(response.content());
    }
}
```
---

I looked at 0.25.0 code version, extract && remark :
```java
public static AiMessage aiMessageFrom(ChatCompletionResponse response) {
    AssistantMessage assistantMessage = response.choices().get(0).message();

    // remark: this content if not null, the function never trigger
    // response message: {"role":"assistant","content":"get_weather\n ```python\ntool_call(city_name='北京')\n```","name":null,"function_call":{"name":"get_weather","arguments":"{\"city_name\": \"北京\"}"}}
    String content = assistantMessage.content();
    if (content != null) {
        return aiMessage(content);
    }

    List<ToolCall> toolCalls = assistantMessage.toolCalls();
    if (toolCalls != null) {
        List<ToolExecutionRequest> toolExecutionRequests = toolCalls.stream()
                .filter(toolCall -> toolCall.type() == FUNCTION)
                .map(InternalOpenAiHelper::toToolExecutionRequest)
                .collect(toList());
        return aiMessage(toolExecutionRequests);
    }

    FunctionCall functionCall = assistantMessage.functionCall();
    if (functionCall != null) {
        ToolExecutionRequest toolExecutionRequest = ToolExecutionRequest.builder()
                .name(functionCall.name())
                .arguments(functionCall.arguments())
                .build();
        return aiMessage(toolExecutionRequest);
    }

    throw illegalArgument("Unexpected response: " + response);
}
```

**Comment by nn200433** [comment]
> this was [fixed](https://github.com/langchain4j/langchain4j/blob/38049a197bac084b9f980a7edaf9619e0115403d/langchain4j-open-ai/src/main/java/dev/langchain4j/model/openai/InternalOpenAiHelper.java#L53), will be available in 0.25.0

I looked someone pr the chat-glm module [#360](https://github.com/langchain4j/langchain4j/pull/360)，I can learn from it.

**Comment by nn200433** [comment]
maybe in calling function also exist problem, For example:
```log
==========================================================================================
[用户]: 帮我查下
预定编号：123-457
姓名：Klaus Heisler
==========================================================================================
2023-12-19 10:30:06.067 DEBUG 7312 --- [           main] dev.langchain4j.service.AiServices       : No relevant information was found
2023-12-19 10:30:06.512 DEBUG 18600 --- [           main] d.a.openai4j.RequestLoggingInterceptor   : Request:
- method: POST
- url: http://172.16.30.34:8000/v1/chat/completions
- headers: [Authorization: Bearer ignored]
- body: {
  "model": "chatglm3-6b",
  "messages": [
    {
      "role": "system",
      "content": "你是一家名为 \"微笑万里行 \"的汽车租赁公司的客户支持顾问。\n在提供有关预订或取消预订的信息之前，您必须始终进行检查：\n预订号码、客户姓名。\n今天是 2023-12-19 。"
    },
    {
      "role": "user",
      "content": "嗨，我忘记了我的预订时间。\n\nHere is some information that might be useful for answering:\n\n6.责任\n6.1 用户须对租赁期间发生的任何损坏、损失或盗窃负责。\n6.2 我们不承担任何间接或后果性损失、损害或费用，包括但不限于利润损失。\n\n7.适用法律\n这些条款受美国法律管辖，并根据美国法律进行解释，与这些条款有关的任何争议均受美国法院的专属管辖。\n\n8.条款变更\n我们可随时通过修改本页面来修订这些使用条款。您应不时查看本页面，以了解我们所做的任何更改。\n\n9.接受本条款\n使用服务即表示您已阅读并理解这些条款，并同意受其约束。\n如果您不同意这些条款，请不要使用或访问我们的服务。"
    },
    {
      "role": "assistant",
      "content": "好的，为了更好地帮助您，我需要确认您的预订信息。请问您的预订号是多少？"
    },
    {
      "role": "user",
      "content": "帮我查下\n预定编号：123-457\n姓名：Klaus Heisler\n\nHere is some information that might be useful for answering:\n\n微笑万里行租车服务使用条款\n\n1.简介\n本服务条款（以下简称 \"条款\"）对您（个人）在世界任何国家访问或使用由 Miles of Smiles Car Rental Services（一家在美国注册的公司）提供的应用程序、网站、内容、产品和服务（以下简称 \"服务\"）进行管理。\n\n2.2. 服务\nMiles of Smiles向最终用户出租车辆。我们保留在任何时候暂时或永久中止服务的权利，并且不对服务的任何修改、暂停或中止负责。\n\n3.预订\n3.1 用户可以通过我们的网站或移动应用程序进行预订。\n3.2 在预订过程中，您必须提供准确、最新和完整的信息。您应对您账户下产生的所有费用负责。\n3.3 所有预订均视车辆供应情况而定。\n\n4.取消政策\n4.1 预订可在预订期开始前 7 天取消。\n4.2 如果预订期少于 3 天，则不允许取消预订。\n\n5.车辆使用\n5.1 从 \"微笑万里行 \"租用的所有车辆不得用于以下用途\n用于任何非法目的或与任何刑事犯罪有关。\n用于教授他人驾驶。\n参加任何比赛、拉力赛或竞赛。\n在酒精或药物影响下使用。"
    }
  ],
  "temperature": 0.8,
  "top_p": 0.8,
  "max_tokens": 100,
  "functions": [
    {
      "name": "cancelBooking",
      "description": "取消预定",
      "parameters": {
        "type": "object",
        "properties": {
          "customerSurname": {
            "description": "需要查询的客户姓氏",
            "type": "string"
          },
          "customerName": {
            "description": "需要查询的客户名字",
            "type": "string"
          },
          "bookingNumber": {
            "description": "需要查询的订单预定编号",
            "type": "string"
          }
        },
        "required": [
          "bookingNumber",
          "customerName",
          "customerSurname"
        ]
      }
    },
    {
      "name": "getBookingDetails",
      "description": "根据订单编号、用户姓名来获取订单详情",
      "parameters": {
        "type": "object",
        "properties": {
          "customerSurname": {
            "description": "需要查询的客户姓氏",
            "type": "string"
          },
          "customerName": {
            "description": "需要查询的客户名字",
            "type": "string"
          },
          "bookingNumber": {
            "description": "需要查询的订单预定编号",
            "type": "string"
          }
        },
        "required": [
          "bookingNumber",
          "customerName",
          "customerSurname"
        ]
      }
    }
  ]
}
2023-12-19 10:30:09.000 DEBUG 18600 --- [           main] d.a.openai4j.ResponseLoggingInterceptor  : Response:
- status code: 200
- headers: [date: Tue, 19 Dec 2023 02:30:05 GMT], [server: uvicorn], [content-length: 411], [content-type: application/json]
- body: {"model":"chatglm3-6b","object":"chat.completion","choices":[{"index":0,"message":{"role":"assistant","content":"好的，我已经为您查询到预订信息。预订号是123-457，您的名字是Klaus Heisler。请问有什么我可以帮您做的吗？","name":null,"function_call":null},"finish_reason":"stop"}],"created":1702953010,"usage":{"prompt_tokens":976,"total_tokens":1018,"completion_tokens":42}}
==========================================================================================
[顾问]: 好的，我已经为您查询到预订信息。预订号是123-457，您的名字是Klaus Heisler。请问有什么我可以帮您做的吗？
==========================================================================================
```

---

description:

normally speaking, this json can be call `getBookingDetails` function
```json
{"role":"user","content":"帮我查下\n预定编号：123-457\n姓名：Klaus Heisler\n\nHere is some information that might be useful for answering:\n\n微笑万里行租车服务使用条款\n\n1.简介\n本服务条款（以下简称 \"条款\"）对您（个人）在世界任何国家访问或使用由 Miles of Smiles Car Rental Services（一家在美国注册的公司）提供的应用程序、网站、内容、产品和服务（以下简称 \"服务\"）进行管理。\n\n2.2. 服务\nMiles of Smiles向最终用户出租车辆。我们保留在任何时候暂时或永久中止服务的权利，并且不对服务的任何修改、暂停或中止负责。\n\n3.预订\n3.1 用户可以通过我们的网站或移动应用程序进行预订。\n3.2 在预订过程中，您必须提供准确、最新和完整的信息。您应对您账户下产生的所有费用负责。\n3.3 所有预订均视车辆供应情况而定。\n\n4.取消政策\n4.1 预订可在预订期开始前 7 天取消。\n4.2 如果预订期少于 3 天，则不允许取消预订。\n\n5.车辆使用\n5.1 从 \"微笑万里行 \"租用的所有车辆不得用于以下用途\n用于任何非法目的或与任何刑事犯罪有关。\n用于教授他人驾驶。\n参加任何比赛、拉力赛或竞赛。\n在酒精或药物影响下使用。"}
```
translate:
```
zh-cn: 帮我查下\n预定编号：123-457\n姓名：Klaus Heisler
en: Check for me\n bookingNumber: 123-457\n Name: Klaus Heisler
```
so, should be called `getBookingDetails` function, but chat-glm unrecognized. I'm try delete Retrieval Content ( `Here is some information that might be useful for answering:xxxxxxx......`) after, the chat-glm can be choose function tools.  For example:
```log
==========================================================================================
[用户]: 帮我查下
预定编号：123-457
姓名：Klaus Heisler
==========================================================================================
2023-12-19 10:56:39.067 DEBUG 7312 --- [           main] dev.langchain4j.service.AiServices       : No relevant information was found
2023-12-19 10:56:39.069 DEBUG 7312 --- [           main] d.a.openai4j.RequestLoggingInterceptor   : Request:
- method: POST
- url: http://172.16.30.34:8000/v1/chat/completions
- headers: [Authorization: Bearer ignored]
- body: {
  "model": "chatglm3-6b",
  "messages": [
    {
      "role": "system",
      "content": "你是一家名为 \"微笑万里行 \"的汽车租赁公司的客户支持顾问。\n在提供有关预订或取消预订的信息之前，您必须始终进行检查：\n预订号码、客户姓名。\n今天是 2023-12-19 。"
    },
    {
      "role": "user",
      "content": "嗨，我忘记了我的预订时间。\n\nHere is some information that might be useful for answering:\n\n6.责任\n6.1 用户须对租赁期间发生的任何损坏、损失或盗窃负责。\n6.2 我们不承担任何间接或后果性损失、损害或费用，包括但不限于利润损失。\n\n7.适用法律\n这些条款受美国法律管辖，并根据美国法律进行解释，与这些条款有关的任何争议均受美国法院的专属管辖。\n\n8.条款变更\n我们可随时通过修改本页面来修订这些使用条款。您应不时查看本页面，以了解我们所做的任何更改。\n\n9.接受本条款\n使用服务即表示您已阅读并理解这些条款，并同意受其约束。\n如果您不同意这些条款，请不要使用或访问我们的服务。"
    },
    {
      "role": "assistant",
      "content": "好的，为了更好地帮助您，我需要确认您的预订信息。请告诉我您的预订号、客户姓名和姓氏。"
    },
    {
      "role": "user",
      "content": "帮我查下\n预定编号：123-457\n姓名：Klaus Heisler"
    }
  ],
  "temperature": 0.8,
  "top_p": 0.8,
  "max_tokens": 100,
  "functions": [
    {
      "name": "cancelBooking",
      "description": "取消预定",
      "parameters": {
        "type": "object",
        "properties": {
          "customerSurname": {
            "description": "需要查询的客户姓氏",
            "type": "string"
          },
          "customerName": {
            "description": "需要查询的客户名字",
            "type": "string"
          },
          "bookingNumber": {
            "description": "需要查询的订单预定编号",
            "type": "string"
          }
        },
        "required": [
          "bookingNumber",
          "customerName",
          "customerSurname"
        ]
      }
    },
    {
      "name": "getBookingDetails",
      "description": "根据订单编号、用户姓名来获取订单详情",
      "parameters": {
        "type": "object",
        "properties": {
          "customerSurname": {
            "description": "需要查询的客户姓氏",
            "type": "string"
          },
          "customerName": {
            "description": "需要查询的客户名字",
            "type": "string"
          },
          "bookingNumber": {
            "description": "需要查询的订单预定编号",
            "type": "string"
          }
        },
        "required": [
          "bookingNumber",
          "customerName",
          "customerSurname"
        ]
      }
    }
  ]
}
2023-12-19 10:56:41.290 DEBUG 7312 --- [           main] d.a.openai4j.ResponseLoggingInterceptor  : Response:
- status code: 200
- headers: [date: Tue, 19 Dec 2023 02:56:35 GMT], [server: uvicorn], [content-length: 484], [content-type: application/json]
- body: {"model":"chatglm3-6b","object":"chat.completion","choices":[{"index":0,"message":{"role":"assistant","content":"getBookingDetails\n ```python\ntool_call(bookingNumber='123-457', customerName='Klaus Heisler')\n```","name":null,"function_call":{"name":"getBookingDetails","arguments":"{\"bookingNumber\": \"123-457\", \"customerName\": \"Klaus Heisler\"}"}},"finish_reason":"function_call"}],"created":1702954601,"usage":{"prompt_tokens":695,"total_tokens":734,"completion_tokens":39}}
2023-12-19 10:56:41.292 DEBUG 7312 --- [           main] d.l.agent.tool.DefaultToolExecutor       : About to execute ToolExecutionRequest { name = "getBookingDetails", arguments = "{"bookingNumber": "123-457", "customerName": "Klaus Heisler"}" } for memoryId default
==========================================================================================
[工具]: 获取预订详情 123-457 for Klaus Heisler null...
==========================================================================================
2023-12-19 10:56:41.303 DEBUG 7312 --- [           main] d.l.agent.tool.DefaultToolExecutor       : Tool execution result: {
  "bookingNumber": "123-457",
  "customer": {
    "name": "Klaus Heisler"
  }
}
```
---
summarize, I think `dev.langchain4j.retriever.Retriever#findRelevant` can be add a judgment. For example:
```java
@Override
public List<TextSegment> findRelevant(String text) {
    // remarks: add a judgment
    // it's my brute force solutions: user query problem contains any string to return empty list
    if (StrUtil.containsAny(text, "帮我查下", "查询", "追踪")) {
        return Collections.emptyList();
    }
    List<EmbeddingMatch<TextSegment>> relevant;
    final Embedding                   embeddedText = embeddingModel.embed(text).content();
    if (minScore == null) {
        relevant = embeddingStore.findRelevant(embeddedText, maxResults);
    } else {
        relevant = embeddingStore.findRelevant(embeddedText, maxResults, minScore);
    }
    return relevant.stream().map(EmbeddingMatch::embedded).collect(toList());
}
```
